FullClusterRestartIT#testSearch fails reproducibly.

**Build scan**: https://gradle-enterprise.elastic.co/s/ip4a5jts77e7e

**Repro line**: 

```
./gradlew ':x-pack:qa:full-cluster-restart:v6.0.0#bwcTest' \
  -Dtests.class="org.elasticsearch.xpack.restart.CoreFullClusterRestartIT" -Dtests.method="testSearch" \
  -Dtests.seed=753FBA64CA26166D -Dtests.security.manager=true -Dtests.locale=ar-SY \
  -Dtests.timezone=America/Menominee -Dcompiler.java=14 -Druntime.java=8
```

**Reproduces locally?**: yes

**Applicable branches**: 7.x

**Failure history**: [Kibana build stats link](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(branch),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'FullClusterRestartIT+AND+testSearch'),sort:!(time,desc))).

**Failure excerpt**:
```
java.lang.AssertionError: {_shards={total=1, failed=0, successful=1, skipped=0}, hits={hits=[{_index=testsearch, _type=doc, _source={field.with.dots=KQEHxkeINo, string=JwHPeqNPqT, bool=true, binary=342ylP////+oyvUIAAAAAA==, float=0.7119018, int=37}, _id=141, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=AZDJQTPFBm, string=MopPTBcGAV, bool=true, binary=342ylP////+oyvUIAAAAAA==, float=0.29160386, int=14}, _id=142, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=AHqCZtzcIs, string=orUUWkLVDI, bool=true, binary=342ylP////+oyvUIAAAAAA==, float=0.42685354, int=14}, _id=143, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=oeIQFRQmtS, string=tcxiWvvYrJ, bool=false, binary=342ylP////+oyvUIAAAAAA==, float=0.9378728, int=35}, _id=144, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=PEeErwoWJs, string=jKLPjeVydq, bool=true, binary=342ylP////+oyvUIAAAAAA==, float=0.845978, int=67}, _id=145, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=VtrhkaPJnf, string=DnmcIlXXNT, bool=true, binary=342ylP////+oyvUIAAAAAA==, float=0.25806874, int=63}, _id=146, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=pWKQbHpzWK, string=URTPFGTLtb, bool=false, binary=342ylP////+oyvUIAAAAAA==, float=0.93169653, int=72}, _id=147, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=GYZJRqJESl, string=nNAcKwCQQr, bool=false, binary=342ylP////+oyvUIAAAAAA==, float=0.83365434, int=33}, _id=148, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=SVIxOVofJW, string=JTCrBzqnQN, bool=false, binary=342ylP////+oyvUIAAAAAA==, float=0.26747727, int=40}, _id=149, _score=1.0}, {_index=testsearch, _type=doc, _source={field.with.dots=kAaWoERTrP, string=GfyVdYpxIC, bool=false, binary=342ylP////+oyvUIAAAAAA==, float=0.5007805, int=3}, _id=150, _score=1.0}], total={value=2282, relation=eq}, max_score=1.0}, took=11, timed_out=false} expected:<2284> but was:<2282>
        at __randomizedtesting.SeedInfo.seed([CD76E3D47A62746E:5FCF76813E7B2CC1]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.elasticsearch.upgrades.AbstractFullClusterRestartTestCase.assertTotalHits(AbstractFullClusterRestartTestCase.java:132)
        at org.elasticsearch.upgrades.FullClusterRestartIT.assertBasicSearchWorks(FullClusterRestartIT.java:536)
        at org.elasticsearch.upgrades.FullClusterRestartIT.testSearch(FullClusterRestartIT.java:170)
```

--- a/qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java
+++ b/qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java
@@ -98,7 +98,6 @@ public class FullClusterRestartIT extends AbstractFullClusterRestartTestCase {
         type = getOldClusterVersion().before(Version.V_6_7_0) ? "doc" : "_doc";
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/57245")
     public void testSearch() throws Exception {
         int count;
         if (isRunningAgainstOldCluster()) {

--- a/server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java
@@ -235,9 +235,7 @@ public class FieldNamesFieldMapper extends MetadataFieldMapper {
             }
             for (String path : paths) {
                 for (String fieldName : extractFieldNames(path)) {
-                    if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
-                        document.add(new Field(fieldType().name(), fieldName, fieldType));
-                    }
+                    document.add(new Field(fieldType().name(), fieldName, Defaults.FIELD_TYPE));
                 }
             }
         }

